receiver/googlecloudspannerreceiver/internal/filter/itemcardinality.go

```diff
@@ -0,0 +1,183 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filter
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/ReneKroon/ttlcache/v2"
+	"go.uber.org/zap"
+)
+
+type Item struct {
+	SeriesKey string
+	Timestamp time.Time
+}
+
+type ItemFilter interface {
+	Filter(source []*Item) ([]*Item, error)
+	Shutdown() error
+	TotalLimit() int
+	LimitByTimestamp() int
+}
+
+type ItemFilterResolver interface {
+	Resolve(metricFullName string) (ItemFilter, error)
+	Shutdown() error
+}
+
+type itemCardinalityFilter struct {
+	metricName         string
+	totalLimit         int
+	limitByTimestamp   int
+	itemActivityPeriod time.Duration
+	logger             *zap.Logger
+	cache              *ttlcache.Cache
+}
+
+type currentLimitByTimestamp struct {
+	limitByTimestamp int
+}
+
+func (f *currentLimitByTimestamp) dec() {
+	f.limitByTimestamp--
```

Comment:
Does this need to have some concurrency protection here?

Is it worth using the atomic library here?

Reply:
If you think it is needed, the other methods will need to protect access with the atomic lib as well.

Reply:
Not for now, but overall idea sounds good. I think it can be done in scope of future improvement (moving cardinality to separate component: processor etc.)